### PR TITLE
Switch diffusionmap iteration to _sliced_trajectory

### LINKED ISSUE #4432: In old `AnalysisBase` subclasses, `self._trajectory` is sometimes used instead of `self._sliced_trajectory`
## Expected behavior ##
The discussion is here: https://github.com/MDAnalysis/mdanalysis/discussions/4426#discussioncomment-8153049

The conclusion was to fix this behavior in following subclasses (based on `git grep` below) in `MDAnalysis/analysis`:

- ~~bat.py~~
- diffusionmap.py
- ~~hole2/hole.py~~
- ~~hydrogenbonds/hbond_analysis.py~~
- ~~hydrogenbonds/hbond_autocorrel.py~~
- ~~waterdynamics.py~~

EDIT: removed those that are obsolete (like `waterdynamics` and `hole2`, or don't actually iterate in a wrong way (like `bat`, `hbond_analysis`)

### PR DESCRIPTION
**Pull request by marinegor**
Fixes #4432

Changes made in this Pull Request:
 - changed iteration pattern in `analysis/diffusionmap.py` to adhere with `frames` argument in `AnalysisBase.run()`

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

----
📚 Documentation preview 📚: https://mdanalysis--4433.org.readthedocs.build/en/4433/

### CODE REVIEW

**Review by hmacdope — APPROVED** [review]
This looks good to me @marinegor! Thanks for following this up.

File: package/CHANGELOG [hunk 1]
```diff
@@ -19,6 +19,8 @@ The rules for this file:
  * 2.8.0
 
 Fixes
+ * Fix `analysis/diffusionmap.py` iteration through trajectory to iteration
+ over `self._sliced_trajectory`, hence supporting `DistanceMatrix.run(frames=...)`
```
Comment (IAlibay) [inline package/CHANGELOG]:
```suggestion
  over `self._sliced_trajectory`, hence supporting `DistanceMatrix.run(frames=...)`
  (PR #4433)
```

**Review by IAlibay — APPROVED** [review]
just the one tiny thing, I'll merge in myself to save time

File: package/CHANGELOG (hunk 1 above)
Comment (IAlibay) [reply package/CHANGELOG]: Also added @marinegor to the changelog entry authors

**Pull request merged**